### DISCUSSION TITLE
How are we meant to handle creating different deployments for the same set of components?

### DISCUSSION THREAD

**Discussion by argargreg** [nasa/fprime#1910] [Q&A] [answered]
Dear all,

We are currently in the process of populating components for our flight software, but some of them would require some limited deployments to be made for demonstration purposes - and at the moment, we swap instances.fpp and topology.fpp from the Top folder in and out depending on which demonstration we want to run. 

I doubt this is the correct way of having a panel of deployments and selecting which to run. What is the guideline to accomplish this? I was expecting to see subfolders in the "Top" folder, each defining a deployment with its own instances.fpp and topology.fpp, such that the one to deploy could be selected using fprime-util build [deployment] (if none are given, the root deployment would have been selected).

Thank you in advance,
Kind regards

**Comment by timcanham** [comment, marked as answer]
It is typical for our projects that we simply define multiple deployments, each one tailored for the test/demo environment we wish to build.

**Reply by argargreg** [reply]
Thanks for the answer. Could you recommend an example of file structure to have a panel of applications sharing the same components+the F' framework or confirm the following (which I'm guessing)? 
/
+-- Flight software
+-- Demo1
+-- Demo2
+-- Components
+-- fprime
Which would require the applications' CMakeLists.txt file to be updated such that the add_fprimesubdirectory() directives refer to one level up.

**Reply by timcanham** [reply]
Your example looks pretty good. What I typically do is:

```
/
+-- Flight
     +--- FlightComponent
+-- Demo1
     +--- Demo1Component
+--Common
     +--- Component1
+--Fprime (as a sub-repo)
```

The CMake file for each deployment, as you said, would just include the directories you want for each deployment. You can use some CMake variables to avoid excessive backslashes. A helpful one I use a lot is: `${FPRIME_PROJECT_ROOT}`. That points to the root of your repo. For example, for your `Flight` deployment, your `CMakeLists.txt` file could have:

```
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/FlightComponent")
add_fprime_subdirectory("${FPRIME_PROJECT_ROOT}/Common/Component1")
```

`${CMAKE_CURRENT_LIST_DIR}` points to the directory where the current `CMakeLists.txt` exists.

**Comment by timcanham** [comment]
The build system allows more than one deployment. You aren't required to use `Ref` and hack it each time.

**Comment by bocchino** [comment]
The FPP modeling is designed so that you can define one set of instances and have different topologies refer to them as needed. You don't have to redefine a new set of instances for each topology. You can also have topologies import other topologies. See the FPP User's Guide for details.
